streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java

```diff
@@ -1088,17 +1036,21 @@ private void writeInputData(final List<KeyValue<Long, Long>> records) {
 
     private void verifyStateStore(final KafkaStreams streams,
                                   final Set<KeyValue<Long, Long>> expectedStoreContent,
-                                  final String reason) throws Exception {
-        final ReadOnlyKeyValueStore<Long, Long> store = IntegrationTestUtils
-            .getStore(300_000L, storeName, streams, QueryableStoreTypes.keyValueStore());
-        assertNotNull(store);
-
-        try (final KeyValueIterator<Long, Long> it = store.all()) {
```

Comment:
noob question: when does `org.apache.kafka.streams.errors.InvalidStateStoreException: The state store, store, may have migrated to another instance.` happen?

Reply:
It happens inside the `all()` function (this is part of the IQv1 API), in which we'd check if the state is still available. So in a word, there are two places where we could throw:

```
store = streams.getStore(); // we could throw here
store.get(); store.all(); // we could also throw here
```

which is quite confusing to users, since they'd expect once the `store` is returned successfully, nothing else exceptional should happen.

IQv2 fixed this issue by combining the "get store" and "query the returned store" into a single call, and hence we can just retry-backoff that single call.

Reply:
Mostly due to rebalances; but in this test case, mostly for those "phantom rebalances" where the hb request is not handled in time causing the instance to think they are out of the group, drop all the partitions (and hence tasks), and re-join.